src/pip/_internal/distributions/sdist.py

```diff
@@ -29,7 +29,7 @@ def prepare_distribution_metadata(self, finder, build_isolation):
         self.req.load_pyproject_toml()
 
         # Set up the build isolation, if this requirement should be isolated
-        should_isolate = self.req.use_pep517 and build_isolation
+        should_isolate = self.req.use_pep517 and build_isolation and not self.req.editable
```

Comment:
This will break some other use-cases...

Reply:
I guess the bug is somewhere higher up and that `build_isolation` should be `False` here. (It's also possible to use `self.req.user_supplied`.)

Reply:
Yeah, I think the bug should be fixed somewhere higher in the call stack. `build_isolation` should not be `True` when the requirement is editable.

Reply:
What use case does this break though? I feel this should work as well.

Reply:
[This](https://pip.pypa.io/en/stable/development/getting-started/#running-pip-from-source-tree) broke. Maybe it would be ok to add `and not self.req.user_supplied` instead, but I'll try find out why `build_isolation` is set to the wrong value instead.

Reply:
Do you know why it breaks? I wouldn’t expect that since it’s conceptually not different from any other editable installs.

Reply:
No, it looks like this:

First on `main`:
```
> git status
On branch main
Your branch is up to date with 'origin/main'
> which python3
/usr/bin/python3
> python3 --version
Python 3.7.3
> python3 -m pip install .
Defaulting to user installation because normal site-packages is not writeable
...
Successfully installed pip-21.2.dev0
> python3 -m venv venv
> source venv/bin/activate
(venv) > python -m pip install
(venv) > python -m pip install -e .
Looking in indexes: https://pypi.org/simple
Obtaining file:///home/simonsm/pip
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 21.2.dev0
    Uninstalling pip-21.2.dev0:
      Successfully uninstalled pip-21.2.dev0
  Running setup.py develop for pip
Successfully installed pip-21.2.dev0
(venv) >
```

Then with my change:
```
(venv) > deactivate
> rm -fr venv
> git checkout -
Switched to branch 'fix-editable-pep-517'
Your branch is up to date with 'origin/fix-editable-pep-517'.
> python3 -m pip install .
Defaulting to user installation because normal site-packages is not writeable
...
Successfully installed pip-21.2.dev0
> python3 -m venv venv
> source venv/bin/activate
(venv) > python -m pip install -e .
Looking in indexes: https://pypi.org/simple
Obtaining file:///home/simonsm/pip
  Installing build dependencies ... done
Installing collected packages: pip
  Found existing installation: pip 18.1
    Not uninstalling pip at /home/<user>/pip/venv/lib/python3.7/site-packages, outside environment /home/<user>/pip/venv
    Can't uninstall 'pip'. No files were found to uninstall.
  Running setup.py develop for pip
Successfully installed pip-18.1
(venv) >
```

So pip doesn't pick up the user installation of pip in the venv with this change.

Reply:
Interesting. The message doesn’t even make sense (the path is inside the environment?) so I’d probably look into why that error is triggered.

Reply:
The problem seems to be that the default value for `build_isolation` is `True`, so this works (on `main`):
```
python3 -m pip install --no-build-isolation -e <path>
```
I guess the default value should still be `False` when `-e` is provided though.